File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example_py/h2x/high_level/realtime_move.py
#!/usr/bin/env python3
#!
#!注意事项
#!
#!1. 机器人模式切换需要确保当前状态安全
#!2. 移动命令需要在合适的模式下执行
#!3. 所有操作结果通过result字段返回，0表示成功，非0表示错误


import humanoid_sdk_py
from humanoid_sdk_py import h2x
import time


def main():
    print(
        "=== h2x 实时运动控制示例 ,仅支持在half_walk_mode模式下控制上半身，下半身暂时不要控制 ==="
    )

    a = 0
    high_level = h2x.H2xHighLevel()
    ret, robot_mode = high_level.GetRobotMode()
    if ret != 0 or robot_mode != 10:
        print("当前未处于半蹲行走模式，错误码：", ret, "当前模式：", robot_mode)
        return ret if ret != 0 else -1

    times = 0
    cmd = h2x.RealtimeCmd()
    cmd.time = 0.1
    while True:
        # Upper body: waist (12-14), left arm (15-21), right arm (22-28)
        # Hold at default position with moderate gains.
        for i in range(12, 29):
            cmd.motor_cmd[i].mode = 0
            cmd.motor_cmd[i].q = 0
            # Waist roll/pitch get higher stiffness for stability
            if i == 13 or i == 14:
                cmd.motor_cmd[i].mode = 1
                cmd.motor_cmd[i].kp = 100.0
            else:
                cmd.motor_cmd[i].kp = 50.0
            cmd.motor_cmd[i].kd = 2.0
        # 腰部转0.5弧度
        if times < 50:
            cmd.motor_cmd[15].mode = 1
            cmd.motor_cmd[15].kp = 100.0
            cmd.motor_cmd[15].kd = 5.0
            a += 0.01
            cmd.motor_cmd[15].q = a
        elif times < 100:
            a -= 0.01
            cmd.motor_cmd[15].mode = 1
            cmd.motor_cmd[15].kp = 100.0
            cmd.motor_cmd[15].kd = 5.0
            cmd.motor_cmd[15].q = a
        else:
            cmd.motor_cmd[15].mode = 1
            cmd.motor_cmd[15].kp = 100.0
            cmd.motor_cmd[15].kd = 5.0
            cmd.motor_cmd[15].q = a
            times = 0
        print(f"pos: {cmd.motor_cmd[15].q}")
        ret = high_level.realtimeMove(cmd)
        if ret != 0:
            print(f"实时运动控制失败, 错误码：{ret}")
            return -1

        time.sleep(0.2)
        times += 1

    print("实时运动控制完成")
    return 0


if __name__ == "__main__":
    exit(main())

File: gr00t/Isaac-GR00T/examples/H20/robots/sdk_package/example/h2x/high_level/realtime_move.cpp
/*
 * 注意事项
 *
 * 1. 机器人模式切换需要确保当前状态安全
 * 2. 移动命令需要在合适的模式下执行
 * 3. 所有操作结果通过result字段返回，0表示成功，非0表示错误
 */

#include <chrono>
#include <cstdint>
#include <iostream>
#include <thread>

#include "common/h2x_def.h"
#include "h2x/h2x_high_level.h"

using namespace humanoid;
using namespace humanoid::h2x;

namespace {
constexpr auto kControlPeriod = std::chrono::milliseconds(100);
constexpr double kControlPeriodSec = 0.1;

// motor_cmd[] 使用 0-based index（关节ID - 1）
// 上半身范围：腰部(12-14)，左臂(15-21)，右臂(22-28)，对应关节ID 13-29
constexpr int kUpperBodyStart = 12;  // WAIST_YAW index
constexpr int kUpperBodyEnd = 29;    // R_WRIST_YAW index + 1（exclusive）

// WAIST_ROLL(14) 和 WAIST_PITCH(15) 使用更高的刚度
constexpr int kWaistRollIndex = 13;   // 关节ID 14 = WAIST_ROLL
constexpr int kWaistPitchIndex = 14;  // 关节ID 15 = WAIST_PITCH

constexpr float kHighKp = 100.0f;
constexpr float kNormalKp = 50.0f;
constexpr float kKd = 2.0f;

// 腰部偏转关节（WAIST_YAW，index 12）做摆动
constexpr int kWaistYawIndex = 12;  // 关节ID 13 = WAIST_YAW
}  // namespace

int main() {
  std::cout << "=== h2x 实时运动控制示例 "
               ",仅支持在half_walk_mode模式下控制上半身，下半身暂时不要控制 ==="
            << std::endl;

  H2xHighLevel highLevel;

  int robotMode = 0;
  int ret = highLevel.GetRobotMode(robotMode);
  if (ret != 0 || robotMode != static_cast<int>(RobotMode::HALF_WALK_MODE)) {
    std::cout << "当前未处于半蹲行走模式，错误码：" << ret
              << "，当前模式：" << robotMode << std::endl;
    return ret != 0 ? ret : -1;
  }

  RealtimeCmd cmd{};
  cmd.time = kControlPeriodSec;

  double a = 0.0;
  int times = 0;

  while (true) {
    // 上半身所有关节保持默认位置，施加适度增益
    for (int i = kUpperBodyStart; i < kUpperBodyEnd; ++i) {
      cmd.motor_cmd[i].mode = 0;
      cmd.motor_cmd[i].q = 0.0f;
      // WAIST_ROLL / WAIST_PITCH 使用更高刚度以保持稳定
      if (i == kWaistRollIndex || i == kWaistPitchIndex) {
        cmd.motor_cmd[i].mode = 1;
        cmd.motor_cmd[i].kp = kHighKp;
      } else {
        cmd.motor_cmd[i].kp = kNormalKp;
      }
      cmd.motor_cmd[i].kd = kKd;
    }

    // 腰部偏转关节做 0 → 0.5 rad → 0 的循环摆动
    if (times < 50) {
      a += 0.01;
      cmd.motor_cmd[kWaistYawIndex].mode = 1;
      cmd.motor_cmd[kWaistYawIndex].q = static_cast<float>(a);
    } else if (times < 100) {
      a -= 0.01;
      cmd.motor_cmd[kWaistYawIndex].mode = 1;
      cmd.motor_cmd[kWaistYawIndex].q = static_cast<float>(a);
    } else {
      cmd.motor_cmd[kWaistYawIndex].mode = 1;
      times = 0;
    }

    std::cout << "pos: " << cmd.motor_cmd[kWaistYawIndex].q << std::endl;

    ret = highLevel.realtimeMove(cmd);
    if (ret != 0) {
      std::cerr << "实时运动控制失败, 错误码：" << ret << std::endl;
      return -1;
    }

    std::this_thread::sleep_for(kControlPeriod);
    ++times;
  }

  std::cout << "实时运动控制完成" << std::endl;
  return 0;
}
